// @tensorflow/tfjs-backend-wasm/dist/kernels/_FusedMatMul.js
"use strict";
/**
 * @license
 * Copyright 2019 Google Inc. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tfjs_core_1 = require("@tensorflow/tfjs-core");
var types_1 = require("./types");
var wasmFusedMatMul;
function setup(backend) {
    wasmFusedMatMul = backend.wasm.cwrap('_FusedMatMul', null /* void */, [
        'number',
        'array',
        'number',
        'number',
        'array',
        'number',
        'number',
        'number',
        'number',
        'number',
        'number',
        'number' // out_id
    ]);
}
function fusedBatchMatMul(args) {
    var inputs = args.inputs, backend = args.backend, attrs = args.attrs;
    var a = inputs.a, b = inputs.b, bias = inputs.bias, preluActivationWeights = inputs.preluActivationWeights;
    if (a.dtype !== 'float32' || b.dtype !== 'float32') {
        throw new Error("_FusedMatMul for non non-float32 tensors not yet supported.");
    }
    var transposeA = attrs.transposeA, transposeB = attrs.transposeB, activation = attrs.activation;
    var aId = backend.dataIdMap.get(a.dataId).id;
    var bId = backend.dataIdMap.get(b.dataId).id;
    var biasId = 0;
    if (bias != null) {
        var biasData = backend.dataIdMap.get(bias.dataId);
        if (biasData.shape.length !== 1) {
            throw new Error("_FusedMatMul only supports rank-1 bias but got " +
                ("rank " + biasData.shape.length + "."));
        }
        biasId = biasData.id;
    }
    var preluActivationWeightsId = preluActivationWeights == null ?
        0 :
        backend.dataIdMap.get(preluActivationWeights.dataId).id;
    var fusedActivation = types_1.FusableActivation[activation];
    if (fusedActivation == null) {
        throw new Error(activation + " activation not yet supported for FusedConv2D " +
            "in the wasm backend.");
    }
    var leftDim = transposeA ? a.shape[2] : a.shape[1];
    var rightDim = transposeB ? b.shape[1] : b.shape[2];
    var batchDim = a.shape[0];
    var out = backend.makeOutput([batchDim, leftDim, rightDim], a.dtype);
    var outId = backend.dataIdMap.get(out.dataId).id;
    var aShapeBytes = new Uint8Array(new Int32Array(a.shape).buffer);
    var bShapeBytes = new Uint8Array(new Int32Array(b.shape).buffer);
    wasmFusedMatMul(aId, aShapeBytes, a.shape.length, bId, bShapeBytes, b.shape.length, transposeA, transposeB, fusedActivation, biasId, preluActivationWeightsId, outId);
    return out;
}
tfjs_core_1.registerKernel({
    kernelName: '_FusedMatMul',
    backendName: 'wasm',
    setupFunc: setup,
    kernelFunc: fusedBatchMatMul
});
