// @tensorflow/tfjs-backend-wasm/dist/kernels/Softmax.js
"use strict";
/**
 * @license
 * Copyright 2020 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tfjs_core_1 = require("@tensorflow/tfjs-core");
var wasmFunc;
function setup(backend) {
    wasmFunc = backend.wasm.cwrap('Softmax', null /* void */, [
        'number',
        'number',
        'number',
        'number' // batch
    ]);
}
function softmax(args) {
    var backend = args.backend, logits = args.inputs.logits, dim = args.attrs.dim;
    var xId = backend.dataIdMap.get(logits.dataId).id;
    var out = backend.makeOutput(logits.shape, logits.dtype);
    var outId = backend.dataIdMap.get(out.dataId).id;
    var channels = logits.shape[dim];
    var batch = tfjs_core_1.util.sizeFromShape(logits.shape) / channels;
    // Short-circuit zero-sized tensors.
    if (tfjs_core_1.util.sizeFromShape(out.shape) === 0) {
        return out;
    }
    wasmFunc(xId, outId, channels, batch);
    return out;
}
tfjs_core_1.registerKernel({
    kernelName: 'Softmax',
    backendName: 'wasm',
    setupFunc: setup,
    kernelFunc: softmax
});
